File: models/Product.ts
import mongoose, { Schema, model, models } from "mongoose";

const ProductSchema = new Schema(
    {
        name: { type: String, required: true },
        category: { type: String, required: true }, // Vegetable, Fruit, Grain, etc.
        price: { type: Number, required: true },
        stock: { type: Number, required: true },
        description: { type: String, required: true },
        image: { type: String, required: true },
        farmerId: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },

        location: { type: String }, // Optional, can inherit from farmer
        unit: { type: String, default: "kg" },
        rating: { type: Number, default: 0 },
        numReviews: { type: Number, default: 0 },
        sales: { type: Number, default: 0 },

        createdAt: { type: Date, default: Date.now },
    },
    { timestamps: true }
);

const Product = models.Product || model("Product", ProductSchema);

export default Product;

File: agri1/lib/models/product.dart
class Product {
  final String id;
  final String name;
  final String description;
  final double price;
  final int stock;
  final String farmerId;
  final String farmerName;

  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    required this.stock,
    required this.farmerId,
    required this.farmerName,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      stock: json['stock'] ?? 0,
      farmerId: json['farmerId'] is String ? json['farmerId'] : (json['farmerId']['_id'] ?? ''),
      farmerName: json['farmerId'] is Map ? (json['farmerId']['name'] ?? '') : '',
    );
  }
}
